### ISSUE TITLE
DR | Retrieve full history of failures

### ISSUE THREAD

**Issue by anniebtran** [department-of-veterans-affairs/va.gov-team#81303]
## Value Statement

**_As a_** VA staff member responsible for remediating supplemental claims or board appeals that silently failed
**_I want to_** know the exact nature of each types of transaction failure
**_So that_** I can create a remediation plan to make the Veteran's claim whole. 

---

## Background Context

Follow-up ticket to #80442. Some notes from the investigation ticket are [here](https://docs.google.com/document/d/1GxYoFA-kJRlY1pSOzbNUhHmv3rn7MqVZ1S_ZMzS7zN0/edit)

## Acceptance Criteria

- [ ] History of failures for HLR has been captured and sent to appropriate contact to investigate which submissions need remediation
- [ ] History of failures for SC has been captured and sent to appropriate contact to investigate which submissions need remediation
  - [ ] History of failures for ancillary forms (ITF and 4142) has been captured and sent to appropriate contact to investigate which submissions need remediation
- [ ] History of failures for NOD has been captured and sent to appropriate contact to investigate which submissions need remediation
- [ ] Any limitations on the amount of failure data we can actually retrieve have been identified

## Out of scope

- Any follow-up work that is identified after CorpDB does their investigations

## Open questions

- Who exactly is going to work with the data we're passing on? (Will need to coordinate with Alejandro Mendelioflores and John Gosnell)
- [x] What data formatting should we use so that we are able to hand off the failure data in a useful way?
_(Answered in Shaun's comment below)_

## Tasks
- [x] Identify what information about failures / what format is needed for hand off (i.e. the data format that Alex has used with Thomas' team)
- [x] Collect database information related to suspected errors/failures
- [x] Download logs available in DataDog online archives
- [x] Download logs from AWS Cloudwatch (logs up to October 2022)
- [x] Download DataDog log archives stored in S3 using Athena
- [x] Store data in format most useful for follow-up work. See Shaun's comment below

## Definition of Ready

- [ ] Clear value description
- [ ] Testable acceptance criteria
- [ ] Accessibility added to acceptance criteria
- [ ] Approved designs attached
- [ ] Sample data provided where appropriate
- [ ] Estimated to fit within the sprint
- [ ] Dependencies and blockers linked

## Definition of Done

- [ ] Meets acceptance criteria
- [ ] Passed E2E testing (90% coverage)
- [ ] Passed unit testing (90% coverage)
- [ ] Passed integration testing (if applicable)
- [ ] Code reviewed (internal)
- [ ] Submitted to staging
- [ ] Team approved production verification process
- [ ] Reviewed and approved by product and/or design

**Comment by shaunburdick** [comment]
Here is the recommended format from @tblackwe for use with CorpDB contacts:

### 3 separate file types:
- <data_type>.csv (1 file)
   - Headers: `data_type`,`form_type`,`va_gov_submission_created_at`,`va_gov_submission_id`,`edipi`,`icn`
   - Content:
      - `data_type`: the name we are using to describe what we are running. eg M1A agreed on with Alex
      - `form_type`: the id of the form. eg 21P-530
      - `va_gov_submission_created_at`: the ActiveRecord creation timestamp
      - `va_gov_submission_id`: The ActiveRecord ID
      - `edipi`: Unique Veteran Identifier, we pull it from the Account.lookup_by_user_uuid response
      - `icn`: another Unique Veteran Identifier, we pull it from the Account.lookup_by_user_uuid response

- <data_type>.json (1 file)
```json
{
    "schema_def": "<link to vets-json-schema for form>", 
    "submissions": [
        {
            "id": "<va_gov_submission_id>", 
            "details": "<json dump of form submission>"
        }
    ]
}
```

- <`va_gov_submission_id`>.zip (many files, 1 per claim)
   - Contents:
      - PDF of Claim Submission
      - Any Veteran uploaded documents

**Comment by anniebtran** [comment]
These are included in the doc I linked in the description, but thought I'd post it here for more visibility

### Initial DB counts

#### Records that likely need remediation (Downstream errors)
- Lighthouse `AppealsApi::EvidenceSubmission` records with "error" status: 739
- Lighthouse Form Submissions with "error" status 
  - SC: 0 
  - NOD: 0
  - HLR: 4
 
#### Records that don't have all the expected DB records accounted for, but could be fine because of exceptions like PACT Act / uploading within x number of days
- Missing `AppealSubmission` record (our receipt of the form submission), but has a Lighthouse Form Submission
  - SC: 89
  - NOD: 9
  - HLR: 1741
- Selected "evidence upload" in the form, but missing `AppealsApi::EvidenceSubmission` record(s)
  - SC: 140
  - NOD: 2036

**Comment by anniebtran** [comment]
Link to Sam R's notes about looking up logs in AWS [here](https://dsva.slack.com/archives/C5AGLBNRK/p1714590726085109?thread_ts=1714583924.807999&cid=C5AGLBNRK)

**Comment by anniebtran** [comment]
So far I've gathered information for errors/failures based on database records and the last year's logs in DataDog. Still need to:
- Hand off queries for rehydrating DataDog to Cory
- Gather logs from AWS
- Format all the data for CorpDB investigation

**Comment by saderagsdale** [comment]
Will give queries to Cory today or tomorrow.

**Comment by anniebtran** [comment]
Working on getting AWS logs, but CLI has been hanging. Will work with Thomas on resolving that or finding a workaround. Also looking into the Datadog archives stored in S3 and seeing if Athena could help us query for the logs we need

**Comment by saderagsdale** [comment]
@anniebtran update tasks to reflect whats done / next up.

**Comment by saderagsdale** [comment]
Still working on getting AWS logs (takes a lot of time).

**Comment by anniebtran** [comment]
Continuing to work on getting access to DD logs that have been archived in S3. Seems like platform is going to work out a secure way to grant access (I think?) and they have a ticket for some [discovery/implementation](https://dsva.slack.com/archives/CBU0KDSB1/p1717075438595199?thread_ts=1716475029.512699&cid=CBU0KDSB1)

**Comment by anniebtran** [comment]
Met with Steve, Kyle M (platform), and Cory. Some updates/notes:
- Athena had only just been stood up once we requested access, and is not at the point where they can just set us up with access directly, which is why the process will be more like data handoff vs. getting in there ourselves
- Kyle's going to pull the first 100 or 1000 records/logs and see how much time that buys us — they will be the oldest logs so that we can try to avoid data falling into glacier
- He'll hand that CSV off to us so that we have somewhere to start while they continue to work in parallel
